### PR TITLE
Fix memory leak with push_back

### PR DESCRIPTION
**Pull request by mppf**
First, pushback-no-leak.chpl is an example that Ben Albrecht provided.  This program was leaking memory because the array-as-vector support creates extra space at the end of the array to make future additions not need to reallocate. The problem was that these array elements were being initialized but never freed.

Fixing this issue amounts to adjusting dsiDestroyArr to figure out the size of the allocated array by looking at dataAllocRange in addition to dom.dsiNumIndices. However, in order to have the current dataAllocRange in the call to dsiDestroyArr in the call to dsiReallocate inside e.g.  push_back, I needed to also modify push_back and the other array-as-vec routines to only leave it to dsiReallocate to set dataAllocRange (and do so only after the reallocation has completed.) To do so, I introduced a temporay variable called nextAllocRange.

Passed quickstart testing.
Passed full local testing.

Reviewed by @daviditen - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
@ben-albrecht - note - this fixes your memory leak

**Comment by mppf** [comment]
@daviditen - care to review this fix for the push_back memory leak?

**Comment by daviditen** [comment]
Sure, I'll review.

**Comment by daviditen** [comment]
This looks good.

**Comment by daviditen** [comment]
Thanks for tracking it down BTW.

File: test/arrays/ferguson/pushback-no-leak.chpl [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/*
+  The following code leaks memory
```
Comment (cassella) [inline test/arrays/ferguson/pushback-no-leak.chpl]: Would it be nitpicking to point out that it no longer does?

**Pull request merged**